lib/Transforms/IPO/StructFieldAccessInfo.cpp

```diff
@@ -86,28 +149,63 @@ StructFieldAccessInfo::calculateFieldNumFromGEP(const User *U) const {
   auto *Op = U->getOperand(0);
   // Make sure Operand 0 is a struct type and matches the current struct type of
   // StructFieldAccessInfo
-  assert(Op->getType()->getPointerElementType()->isStructTy() &&
-         Op->getType()->getPointerElementType() == StructureType);
-  if (U->getNumOperands() < 3) // GEP to calculate struct field needs at least 2
-                               // indices (operand 1 and 2)
-    return 0;
-  // Operand 1 should be first index to the struct, usually 0; if not 0, it's
-  // like goto an element of an array of structs
-  Op = U->getOperand(1);
-  // TODO: ignore this index for now because it's the same for an array of
-  // structs
-  assert(Op->getType()->isIntegerTy());
-  // Operand 2 should be the index to the field, and be a constant
-  Op = U->getOperand(2);
-  assert(isa<Constant>(Op));
-  auto *Index = cast<Constant>(Op);
-  auto Offset = (FieldNumType)Index->getUniqueInteger().getZExtValue();
-  assert(Offset < NumElements);
-  // TODO: ignore indices after this one. If there's indices, the field has to
-  // be an array or struct
-  return Offset + 1; // return field number starting from 1
+  if (Op->getType()->getPointerElementType()->isStructTy() &&
+      Op->getType()->getPointerElementType() == StructureType) {
+    if (U->getNumOperands() <
+        3) // GEP to calculate struct field needs at least 2
+           // indices (operand 1 and 2)
+      return 0;
+    // Operand 1 should be first index to the struct, usually 0; if not 0, it's
+    // like goto an element of an array of structs
+    Op = U->getOperand(1);
+    // TODO: ignore this index for now because it's the same for an array of
+    // structs
+    assert(Op->getType()->isIntegerTy());
+    // Operand 2 should be the index to the field, and be a constant
+    Op = U->getOperand(2);
+    assert(isa<Constant>(Op));
+    auto *Index = cast<Constant>(Op);
+    auto Offset = (FieldNumType)Index->getUniqueInteger().getZExtValue();
+    assert(Offset < NumElements);
+    // TODO: ignore indices after this one. If there's indices, the field has to
+    // be an array or struct
+    return Offset + 1; // return field number starting from 1
+  } else if (Op->getType()->getPointerElementType()->isArrayTy() &&
```

Comment:
This makes it hard to read. If you have a really complex condition in if, it is better to move it into a separate static function. May be add a FIXME here.